### PR TITLE
Add Go to Super Implementation to context menu and command palette

### PR DESCRIPTION
**Pull request by jdneo**
demo

Note: so far, if the selected element has no super implementation, we are not able to show the editor message. This need VS Code to support such API. See: https://github.com/microsoft/vscode/issues/95308

Signed-off-by: Sheng Chen <sheche@microsoft.com>

### CODE REVIEW

**Comment by fbricon** [comment]
The fact the menu item shows up **everywhere** is very distracting, given than there's no feedback when you call it

**Comment by jdneo** [comment]
@akaroml Let's sync with VS Code team about this issue. We need the API to show the editor message. Otherwise the `no-feedback` behavior would make user confused.

**Comment by jdneo** [comment]
Update: VS Code team has exposed a new API to show the editor message. I will update this PR after the next VS Code stable version released.

**Comment by jdneo** [comment]
@fbricon @testforstephen PR updated

File: src/standardLanguageClient.ts [hunk 1]
```diff
@@ -208,12 +208,55 @@ export class StandardLanguageClient {
...
+				}
+
+				return commands.executeCommand(
+					'editor.action.goToLocations',
+					window.activeTextEditor.document.uri,
+					window.activeTextEditor.selection.active,
+					[],
+					'goto',
+					'No super implementation found'
+				);
```
Comment (testforstephen) [inline src/standardLanguageClient.ts:259]:
```typescript
CommandsRegistry.registerCommand({
	id: 'editor.action.goToLocations',
	description: {
		description: 'Go to locations from a position in a file',
		args: [
			{ name: 'uri', description: 'The text document in which to start', constraint: URI },
			{ name: 'position', description: 'The position at which to start', constraint: corePosition.Position.isIPosition },
			{ name: 'locations', description: 'An array of locations.', constraint: Array },
			{ name: 'multiple', description: 'Define what to do when having multiple results, either `peek`, `gotoAndPeek`, or `goto' },
			{ name: 'noResultsMessage', description: 'Human readable message that shows when locations is empty.' },
		]
	},
```

Can you also use this function to jump to target location when there is result?

Comment (jdneo) [reply src/standardLanguageClient.ts:259]: Tried but seems not working. I'm asking in VS Code repo...

Comment (jdneo) [reply src/standardLanguageClient.ts]: Finally it turns out that all the types should be generated by their constructors in API. Simply making the structure won't work.

**Comment by fbricon** [comment]
@snjeza can you check this PR on Theia/Che?

**Comment by snjeza** [comment]
>  can you check this PR on Theia/Che?

The Go To Super Implementation command doesn't work on Theia.
I am getting:
```
root ERROR Fail to open 'command:java.action.navigateToSuperImplementation?%5B%7B%22uri%22%3A%22ZmlsZTovLy9ob21lL3NucGUvdGVzdGxzcC9zcHJpbmctcGV0Y2xpbmljL3NyYy9tYWluL2phdmEvb3JnL3NwcmluZ2ZyYW1ld29yay9zYW1wbGVzL3BldGNsaW5pYy9tb2RlbC9CYXNlRW50aXR5LmphdmE%3D%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A41%2C%22character%22%3A19%7D%2C%22end%22%3A%7B%22line%22%3A41%2C%22character%22%3A24%7D%7D%7D%5D': validateConstraint@http://localhost:3000/vs/editor/editor.main.js:16167:19
validateConstraints@http://localhost:3000/vs/editor/editor.main.js:16142:31
exports.CommandsRegistry</class_1.prototype.registerCommand/idOrCommand.handler@http://localhost:3000/vs/editor/editor.main.js:81296:29
InstantiationService</InstantiationService.prototype.invokeFunction@http://localhost:3000/vs/editor/editor.main.js:83001:27
execute@http://localhost:3000/66.bundle.js:690:64
```

**Comment by jdneo** [comment]
Is it caused by the new command introduced in VS Code 1.47.0: `editor.action.goToLocations`?

File: package.json [hunk 2]
```diff
@@ -724,6 +729,11 @@
           "command": "java.projectConfiguration.update",
           "when": "resourceFilename =~ /(.*\\.gradle)|(pom.xml)$/",
           "group": "1_javaactions"
+        },
+        {
+          "command": "java.action.navigateToSuperImplementation",
+          "when": "javaLSReady && editorTextFocus && editorLangId == java",
```
Comment (fbricon) [inline package.json:735]: can we add a supportsGoToSuperImplementation flag, computed dynamically, similar to javaLSReady?

File: package.json [hunk 3]
```diff
@@ -735,6 +745,10 @@
           "command": "java.project.import",
           "when": "javaLSReady"
         },
+        {
+          "command": "java.action.navigateToSuperImplementation",
+          "when": "javaLSReady && editorLangId == java"
```
Comment (fbricon) [inline package.json:750]: also check if `supportsGoToSuperImplementation`

File: src/standardLanguageClient.ts [hunk 4]
```diff
@@ -208,12 +208,50 @@ export class StandardLanguageClient {
 				applyWorkspaceEdit(obj, this.languageClient);
 			}));
 
-			context.subscriptions.push(commands.registerCommand(Commands.NAVIGATE_TO_SUPER_IMPLEMENTATION_COMMAND, (location: any) => {
-				const range = this.languageClient.protocol2CodeConverter.asRange(location.range);
-				window.showTextDocument(Uri.parse(decodeBase64(location.uri)), {
-					preserveFocus: true,
-					selection: range,
-				});
+			context.subscriptions.push(commands.registerCommand(Commands.NAVIGATE_TO_SUPER_IMPLEMENTATION_COMMAND, async (location: LinkLocation | Uri) => {
```
Comment (fbricon) [inline src/standardLanguageClient.ts:211]: only register if `supportsGoToSuperImplementation === true`

**Comment by jdneo** [comment]
@fbricon @snjeza Is Theia/Che support the `goto super implementation` in the hover? I mean, should the flag `supportsGoToSuperImplementation` used to control the entire `goto super implementation` feature, or just some scenarios of it. (for example, only for the context menu and command palette)

**Comment by snjeza** [comment]
> Is Theia/Che support the goto super implementation in the hover? 

It works in Theia without this PR.

**Comment by fbricon** [comment]
> > Is Theia/Che support the goto super implementation in the hover?
> 
> It works in Theia without this PR.

you mean a `supportsGoToSuperImplementation` flag is not required?

**Comment by snjeza** [comment]
> you mean a supportsGoToSuperImplementation flag is not required?

I mean it is required.

**Comment by snjeza** [comment]
@jdneo could you resolve conflicts?

**Comment by jdneo** [comment]
PR updated. Now it's using `window.showTextDocument()` to do the navigation. And the go-to-location command will only be used to show the editor message (no super implementation found).

In Che/Theia, since that command so far does not support to show such message, so nothing will happen when no result found.

An issue has been created in the theia repo: https://github.com/eclipse-theia/theia/issues/8212

**Comment by jdneo** [comment]
I also observed that in my machine, sometimes it will take ~10+s to get the result of super implementation. @fbricon @snjeza @testforstephen @Eskibear Have you met that case?

**Comment by fbricon** [comment]
works fine for me

**Comment by fbricon** [comment]
@jdneo thanks!

**Pull request merged**